File: Optima/ActiveStepper.cpp
// Optima is a C++ library for solving linear and non-linear constrained optimization problems
//
// Copyright (C) 2014-2018 Allan Leal
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

#include "ActiveStepper.hpp"

// C++ includes
#include <cassert>

// Optima includes
#include <Optima/Exception.hpp>
#include <Optima/IndexUtils.hpp>
#include <Optima/SaddlePointMatrix.hpp>
#include <Optima/SaddlePointSolver.hpp>
#include <Optima/Options.hpp>
#include <Optima/Utils.hpp>

namespace Optima {

struct ActiveStepper::Impl
{
    Options options;          ///< The options for the optimization calculation
    Matrix W;                 ///< The coefficient matrix W = [A; J] of the linear/nonlinear equality constraints.
    Vector z;                 ///< The instability measures of the variables defined as `z = g + tr(W)*y`.
    Vector s;                 ///< The solution vector `s = [dx dy]` for the saddle point problem.
    Vector r;                 ///< The right-hand side residual vector `r = [rx ry]`  for the saddle point problem.
    Index n      = 0;         ///< The number of variables in x.
    Index ns     = 0;         ///< The number of stable variables in x.
    Index nu     = 0;         ///< The number of unstable (lower/upper) variables in x.
    Index nlower = 0;         ///< The number of variables with lower bounds.
    Index nupper = 0;         ///< The number of variables with upper bounds.
    Index ml     = 0;         ///< The number of linear equality constraints.
    Index mn     = 0;         ///< The number of non-linear equality constraints.
    Index m      = 0;         ///< The number of equality constraints (m = ml + mn).
    Index t      = 0;         ///< The total number of variables in x and y (t = n + m).
    SaddlePointSolver solver; ///< The saddle point solver.
    Indices ilowerpartition;  ///< The ordering of the variables as (with lower bounds, without lower bounds, fixed)
    Indices iupperpartition;  ///< The ordering of the variables as (with upper bounds, without upper bounds, fixed)

    /// Construct a default ActiveStepper::Impl instance.
    Impl()
    {}

    /// Construct a ActiveStepper::Impl instance.
    Impl(ActiveStepperInitArgs args)
    : n(args.n), m(args.m), W(args.m, args.n)
    {
        // Ensure the step calculator is initialized with a positive number of variables.
        Assert(n > 0, "Could not proceed with ActiveStepper initialization.",
            "The number of variables is zero.");

        // Initialize number of linear and nonlinear equality constraints
        ml = args.A.rows();
        mn = m - ml;

        // Initialize the matrix W = [A; J], with J=0 at this initialization time (updated at each decompose call)
        W << args.A, zeros(mn, n);

        // Initialize total number of variables x and y
        t  = n + m;

        // Initialize auxiliary vectors
        z = zeros(n);
        r = zeros(t);
        s = zeros(t);
    }

    /// Initialize the step calculator before calling decompose multiple times.
    auto initialize(ActiveStepperInitializeArgs args) -> void
    {
        // Auxiliary references
        auto xlower = args.xlower;
        auto xupper = args.xupper;
        auto iordering = args.iordering;

        // Ensure consistent dimensions of vectors/matrices.
        assert(xlower.rows() == n);
        assert(xupper.rows() == n);
        assert(iordering.rows() == n);

        // Initialize the initial ordering of the variables
        iordering = indices(n);

        // Update the ordering of the variables with lower and upper bounds
        auto has_lower_bound = [&](Index i) -> bool { return !std::isinf(xlower[i]); };
        auto has_upper_bound = [&](Index i) -> bool { return !std::isinf(xupper[i]); };

        ilowerpartition = iordering;
        iupperpartition = iordering;

        nlower = moveLeftIf(ilowerpartition, has_lower_bound);
        nupper = moveLeftIf(iupperpartition, has_upper_bound);
    }

    /// Decompose the saddle point matrix for diagonal Hessian matrices.
    auto decompose(ActiveStepperDecomposeArgs args) -> void
    {
        // Ensure the step calculator has been initialized.
        Assert(n != 0, "Could not proceed with ActiveStepper::decompose.",
            "ActiveStepper object not initialized.");

        // Auxiliary references
        auto x         = args.x;
        auto y         = args.y;
        auto g         = args.g;
        auto H         = args.H;
        auto J         = args.J;
        auto xlower    = args.xlower;
        auto xupper    = args.xupper;
        auto iordering = args.iordering;
        auto& nul      = args.nul;
        auto& nuu      = args.nuu;
        const auto A   = W.topRows(ml);

        // Ensure consistent dimensions of vectors/matrices.
        assert(x.rows() == n);
        assert(y.rows() == m);
        assert(g.rows() == n);
        assert(H.rows() == n);
        assert(H.cols() == n);
        assert(J.rows() == mn);
        assert(J.cols() == n || mn == 0);
        assert(A.rows() == ml);
        assert(A.cols() == n || ml == 0);
        assert(iordering.rows() == n);
        assert(xlower.rows() == n);
        assert(xupper.rows() == n);

        // Update the coefficient matrix W = [A; J] with the updated J block
        W.bottomRows(mn) = J;

        // Calculate the optimality residuals
        z.noalias() = g + tr(W)*y;

        // The indices of the variables with lower bounds (ilower) and upper bounds (iupper)
        auto ilower = ilowerpartition.head(nlower);
        auto iupper = iupperpartition.head(nupper);

        // Update the ordering of the variables with lower and upper bounds
        auto is_lower_unstable = [&](Index i) { return x[i] == xlower[i] && z[i] > 0.0; };
        auto is_upper_unstable = [&](Index i) { return x[i] == xupper[i] && z[i] < 0.0; };

        // Organize ilower and iupper as (unstable, stable)
        nul = moveLeftIf(ilower, is_lower_unstable);
        nuu = moveLeftIf(iupper, is_upper_unstable);

        // Update the number of unstable and stable variables
        nu = nul + nuu;
        ns = n - nu;

        // The indices of the lower and upper unstable variables
        // Remember: ilower and iupper are organized in the order [unstable variables, stable variables]!
        auto iul = ilower.head(nul);
        auto iuu = iupper.head(nuu);

        // Move all upper unstable variables to the right in iordering
        moveIntersectionRight(iordering, iuu);

        // Move all lower unstable variables to the right in iordering, but before the unstable variables
        moveIntersectionRight(iordering.head(n - nuu), iul);

        // The indices of the lower and upper unstable variables
        auto iu = iordering.tail(nu);

        // Setup the saddle point matrix.
        // Consider lower/upper unstable variables as "fixed" variables in the saddle point problem.
        // Reason: we do not need to compute Newton steps for the currently unstable variables!
        SaddlePointMatrix spm(H, zeros(n), A, J, iu);

        // Decompose the saddle point matrix (this decomposition is later used in method solve, possibly many times!)
        solver.decompose(spm);
    }

    /// Solve the saddle point problem.
    auto solve(ActiveStepperSolveArgs args) -> void
    {
        // Auxiliary references
        auto x         = args.x;
        auto y         = args.y;
        auto b         = args.b;
        auto g         = args.g;
        auto h         = args.h;
        auto iordering = args.iordering;
        auto dx        = args.dx;
        auto dy        = args.dy;
        auto rx        = args.rx;
        auto ry        = args.ry;
        auto z         = args.z;
        const auto A   = W.topRows(ml);

        // Ensure consistent dimensions of vectors/matrices.
        assert(x.rows() == n);
        assert(y.rows() == m);
        assert(g.rows() == n);
        assert(A.rows() == ml);
        assert(A.cols() == n || ml == 0);
        assert(iordering.rows() == n);

        // The indices of the lower and upper unstable variables
        auto iu = iordering.tail(nu);

        // Calculate the instability measure of the variables.
        z.noalias() = g + tr(W)*y;

        // Calculate the residuals of the first-order optimality conditions
        rx = -z;

        // Set residuals wrt unstable variables to zero
        rx(iu).fill(0.0);

        // Calculate the residuals of the feasibility conditions
        ry.head(ml).noalias() = -(A*x - b);
        ry.tail(mn).noalias() = -h;

        // Solve the saddle point problem
        solver.solve({rx, ry}, {dx, dy});
    }

    /// Compute the sensitivity derivatives of the saddle point problem.
    auto sensitivities(ActiveStepperSensitivitiesArgs args) -> void
    {
        // Auxiliary references
        auto dxdp = args.dxdp;
        auto dydp = args.dydp;
        auto dzdp = args.dzdp;
        auto dgdp = args.dgdp;
        auto dbdp = args.dbdp;
        auto dhdp = args.dhdp;

        // The number of parameters for sensitiviry computation
        auto np = dxdp.cols();

        // Ensure consistent dimensions of vectors/matrices.
        assert(dxdp.rows() == n);
        assert(dydp.rows() == m);
        assert(dzdp.rows() == n);
        assert(dgdp.rows() == n);
        assert(dbdp.rows() == ml);
        assert(dhdp.rows() == mn);
        assert(dydp.cols() == np);
        assert(dzdp.cols() == np);
        assert(dgdp.cols() == np);
        assert(dbdp.cols() == np);
        assert(dhdp.cols() == np);

        // Calculate the residuals of the first-order optimality conditions
        dxdp = -dgdp;

        // Calculate the residuals of the feasibility conditions
        dydp.topRows(ml) = dbdp;
        dydp.bottomRows(mn) = dhdp;

        // Solve the saddle point problem
        for(Index i = 0; i < dxdp.cols(); ++i)
            solver.solve({ dxdp.col(i), dydp.col(i) }, { dxdp.col(i), dydp.col(i) });

        // Calculate the instability measure of the variables.
        dzdp.noalias() = dgdp + tr(W)*dydp;
    }
};

ActiveStepper::ActiveStepper()
: pimpl(new Impl())
{}

ActiveStepper::ActiveStepper(ActiveStepperInitArgs args)
: pimpl(new Impl(args))
{}

ActiveStepper::ActiveStepper(const ActiveStepper& other)
: pimpl(new Impl(*other.pimpl))
{}

ActiveStepper::~ActiveStepper()
{}

auto ActiveStepper::operator=(ActiveStepper other) -> ActiveStepper&
{
    pimpl = std::move(other.pimpl);
    return *this;
}

auto ActiveStepper::setOptions(const Options& options) -> void
{
    pimpl->options = options;
    pimpl->solver.setOptions(options.kkt);
}

auto ActiveStepper::initialize(ActiveStepperInitializeArgs args) -> void
{
    return pimpl->initialize(args);
}

auto ActiveStepper::decompose(ActiveStepperDecomposeArgs args) -> void
{
    return pimpl->decompose(args);
}

auto ActiveStepper::solve(ActiveStepperSolveArgs args) -> void
{
    return pimpl->solve(args);
}

auto ActiveStepper::sensitivities(ActiveStepperSensitivitiesArgs args) -> void
{
    return pimpl->sensitivities(args);
}

} // namespace Optima

File: tests/ActiveStepper.py
# Optima is a C++ library for numerical solution of linear and nonlinear programing problems.
#
# Copyright (C) 2014-2018 Allan Leal
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

from optima import *
from numpy import *
from numpy.linalg import norm
from pytest import approx, mark
from itertools import product

from utils.matrices import testing_matrices_A, matrix_non_singular


from enum import Enum

class MatrixStructure(Enum):
    Zero = 1
    Diagonal = 2
    Dense = 3


# The number of variables
n = 15

# Tested cases for the matrix A
tested_matrices_A = testing_matrices_A

# Tested cases for the structure of matrix H
tested_structures_H = [
    MatrixStructure.Dense,
    MatrixStructure.Diagonal
]

# Tested cases for the indices of fixed variables
tested_ifixed = [
    [],
    [1],
    [1, 3, 7, 9]
]

# Tested cases for the indices of variables with lower bounds
tested_ilower = [
    [],
    [1],
    [1, 3, 7, 9],
    list(range(n))  # all variables with lower bounds
]

# Tested cases for the indices of variables with upper bounds
tested_iupper = [
    [],
    [1],
    [1, 3, 7, 9],
    list(range(n))  # all variables with upper bounds
]

# Tested cases for the saddle point methods
tested_methods = [
    SaddlePointMethod.Fullspace,
    SaddlePointMethod.Nullspace,
    SaddlePointMethod.Rangespace,
]

# Tested number of rows in matrix A and J
tested_ml = [6, 4]
tested_mn = [3, 1, 0]

# Combination of all tested cases
testdata = product(tested_matrices_A,
                   tested_structures_H,
                   tested_ml,
                   tested_mn,
                   tested_ifixed,
                   tested_ilower,
                   tested_iupper,
                   tested_methods)


@mark.parametrize("args", testdata)
def test_active_stepper(args):

    assemble_A, structure_H, ml, mn, ifixed, ilower, iupper, method = args

    # Add the indices of fixed variables to those that have lower and upper bounds
    # since fixed variables are those that have identical lower and upper bounds
    ilower = list(set(ilower + ifixed))
    iupper = list(set(iupper + ifixed))

    # The number variables with finite lower and upper bounds, and fixed variables
    nlower = len(ilower)
    nupper = len(iupper)
    nfixed = len(ifixed)

    # The total number of rows in W = [A; J]
    m = ml + mn

    # The total number of variables x and y
    t = n + m

    # Assemble the coefficient matrix W = [A; J]
    W = assemble_A(m, n, ifixed)

    # Create references to the A and J blocks in W
    A = W[:ml, :]
    J = W[ml:, :]

    # Create vectors for the lower and upper bounds of the variables
    xlower = full(n, -inf)
    xupper = full(n,  inf)

    # Set lower and upper bounds to negative and positive sequence respectively
    xlower[ilower] = -linspace(1.0, nlower, nlower) * 100
    xupper[iupper] =  linspace(1.0, nupper, nupper) * 100

    # Set lower and upper bounds to equal values for fixed variables
    xlower[ifixed] = xupper[ifixed] = linspace(1, nfixed, nfixed) * 10

    # Create vectors x and y
    x = linspace(1.0, n, n)
    y = linspace(1.0, m, m)

    # Create the Hessian matrix *H*
    H = matrix_non_singular(n)

    # Ensure off-diagonal entries are zero if Rangespace method is used
    if method == SaddlePointMethod.Rangespace:
        H = eigen.diag(diag(H))

    # Auxiliary functions to get first and last of a sequence in a set
    first = lambda seq: [seq[0]] if len(seq) > 0 else []
    last = lambda seq: [seq[-1]] if len(seq) > 0 else []

    # Set first and last variables with lower/upper bounds to be unstable
    iunstable_lower = list(set(first(ilower) + last(ilower)))
    iunstable_upper = list(set(first(iupper) + last(iupper)))
    iunstable = list(set(iunstable_lower + iunstable_upper))

    # Create vector s = (dx, dy) with the expected Newton step values
    s = s_expected = linspace(1.0, t, t)  # Introduce `s` as an alias for `s_expected`

    # Get references to the segments dx and dy in s
    dx = s[:n]
    dy = s[n:]

    # Set to zero the entries in dx corresponding to lower/upper unstable variables
    dx[iunstable] = 0.0

    # Set lower/upper unstable variables in x to their respective lower/upper bounds
    x[iunstable_lower] = xlower[iunstable_lower]
    x[iunstable_upper] = xupper[iunstable_upper]

    # Assemble the matrix M = [H, W.T; W, 0] setting rows/cols wrt ifixed to zero
    M = zeros((t, t))
    M[:n, :n] = H
    M[:n, n:] = W.T
    M[n:, :n] = W

    # Set rows wrt to unstable variables to zero and one in the diagonal entry
    M[iunstable, :] = 0.0
    M[iunstable, iunstable] = 1.0

    # Compute the right-hand side vector r = Ms
    r = r_expected = M @ s  # Introduce `r` as an alias for `r_expected`

    # Get references to rx and ry in r where rx := -(g + tr(W)y) and ry := (ryl, ryn)
    rx = r[:n]
    ry = r[n:]

    # Get references to ryl and ryn in ry where ryl := -(Ax - b) and ryn := -h
    ryl = ry[:ml]
    ryn = ry[ml:]

    # Compute the gradient vector remembering that rx := -(g + tr(W)y) for stable variables, zero otherwise
    g = -(rx + W.T @ y)

    # Ensure g has values for lower/upper unstable variables that make these recognized as unstable indeed
    # Note that:
    #  - x[i] is lower unstable when x[i] == xlower[i] and z[i] > 0, where z = g + tr(W)*y
    #  - x[i] is upper unstable when x[i] == xupper[i] and z[i] < 0, where z = g + tr(W)*y
    g[iunstable_lower] =  inf
    g[iunstable_upper] = -inf

    # Compute the right-hand side vector b remembering that ry = -(Ax - b)
    b = ryl + A @ x

    # Compute the residual vector h of the nonlinear equality constraints h(x) = 0 knowing that ryn := -h
    h = -ryn

    # Create the ordering vector that will order the
    # variables as (*stable*, *lower unstable*, *upper unstable*)
    iordering = arange(n)

    # Create the index numbers that will contain the number
    # of *lower unstable* (nul) and *upper unstable* (nuu) variables
    nul = IndexNumber()
    nuu = IndexNumber()

    # The solution of the Newton step calculation
    dx = zeros(n) # The Newton step for the primal variables *x*.
    dy = zeros(m) # The Newton step for the Lagrange multipliers *y*.
    rx = zeros(n) # The residuals of the first-order optimality conditions.
    ry = zeros(m) # The residuals of the linear/nonlinear feasibility conditions.
    z = zeros(n)  # The *unstabilities* of the variables defined as *z = g + tr(W)y* where *W = [A; J]*.

    # Set options for the Newton step calculation
    options = Options()
    options.kkt.method = method

    # Create a Newton step calculator
    stepper = ActiveStepper(n, m, A)
    stepper.setOptions(options)

    # Initialize, decompose the saddle point matrix, and solve the Newton step
    stepper.initialize(xlower, xupper, iordering)
    stepper.decompose(x, y, g, H, J, xlower, xupper, iordering, nul, nuu)
    stepper.solve(x, y, b, h, g, iordering, dx, dy, rx, ry, z)

    # Compare the actual and expected Newton steps
    s_actual = concatenate([dx, dy])

    assert allclose(M @ s_actual, M @ s_expected)

    # Compare the actual and expected right-hand side vector r of residuals
    r_actual = concatenate([rx, ry])

    assert allclose(r_actual, r_expected)

    # Compare the actual and expected z vector
    z_actual = z
    z_expected = g + W.T @ y

    assert allclose(z_actual, z_expected)

    # Compare the actual and expected number of lower/upper unstable variables
    assert nul.value == [x[i] == xlower[i] and z[i] > 0.0 for i in range(n)].count(True)
    assert nuu.value == [x[i] == xupper[i] and z[i] < 0.0 for i in range(n)].count(True)

    set_printoptions(suppress=True, linewidth=1000, precision=3)
    print()
    print(f"assemble_A  = {assemble_A}")
    print(f"structure_H = {structure_H}")
    print(f"ml          = {ml}")
    print(f"mn          = {mn}")
    print(f"ifixed      = {ifixed}")
    print(f"ilower      = {ilower}")
    print(f"iupper      = {iupper}")
    print(f"method      = {method}")
    print()
    print(f"M = \n{M}")
    print(f"dx(actual)   = {dx}")
    print(f"dx(expected) = {s[:n]}")
    print(f"dy(actual)   = {dy}")
    print(f"dy(expected) = {s[n:]}")
    print(f"rx(actual)   = {rx}")
    print(f"rx(expected) = {r[:n]}")
    print(f"ry(actual)   = {ry}")
    print(f"ry(expected) = {r[n:]}")
    print()
